### PR TITLE
test: Mypy next steps

### PR DESCRIPTION
**Pull request by am1r021**
Generate stub files for data and test_framework packages. Set MYPYPATH variable to allow recognition of the aforementioned stub files while ignoring the zmq import that causes an error while running mypy. Fixes issue #19389.

### CODE REVIEW

File: test/functional/test_framework/test_framework.py [hunk 1]
```diff
@@ -714,7 +714,7 @@ def _initialize_chain_clean(self):
     def skip_if_no_py3_zmq(self):
         """Attempt to import the zmq package and skip the test if the import fails."""
         try:
-            import zmq  # noqa
+            import zmq  # type:ignore
```
Comment (adamjonas) [inline test/functional/test_framework/test_framework.py:717]: This line appears to be getting caught in the linter. 

`test/functional/test_framework/test_framework.py:717:13: F401 'zmq' imported but unused`

Comment (troygiorshev) [reply test/functional/test_framework/test_framework.py:717]: As far as I can tell, we don't be able to fix this short of disabling flake8 linting for the entire file.  It's a false positive, in this case the very act of importing the module is "using" it.

**Comment by troygiorshev** [comment]
This all looks great!

However, try as I might, I can't get MyPy to ever actually tell me something is wrong.  (i.e. I intentionally give a function a variable with the wrong type and MyPy still passes every file)  Have you been able to?

Regarding your concern that this isn't python 3.5 compatible, I don't think you have to worry.  The only parts that aren't compatible are the `.pyi` files, but that's ok, they're only used by MyPy.  I've testing running `lint-python.sh` under python 3.5 and it doesn't give any errors.  (MyPy does, however, get confused and tell me that a couple lines are wrong when they aren't.  It doesn't seem related though).

File: test/functional/test_framework/test_framework.py (hunk 1 above)
Comment (troygiorshev) [reply test/functional/test_framework/test_framework.py:717]: It's worth noting that this brings us from 3 `F401` warnings up to 4 `F401` warnings.

Comment (jnewbery) [reply test/functional/test_framework/test_framework.py:717]: is it not possible to keep the `# noqa` linter disable comment?

Comment (troygiorshev) [reply test/functional/test_framework/test_framework.py:717]: Unfortunately not.  We need the `type:ignore` or else MyPy fails completely.  I've tried things like putting them both on the same line but it looks like flake8 and MyPy just don't play nicely together in this sort of conflict.

Worth noting maybe that pylint allows us to disable a particular linting error for an entire function, where AFAICT flake8 does not.

Comment (jnewbery) [reply test/functional/test_framework/test_framework.py:717]: that's unfortunate :(

**Comment by Ghorbanian** [comment]
> This all looks great!
> 
> However, try as I might, I can't get MyPy to ever actually tell me something is wrong. (i.e. I intentionally give a function an argument with the wrong type and MyPy still passes every file) Have you been able to?
> 
> Regarding your concern [here](https://github.com/bitcoin/bitcoin/issues/19389#issuecomment-659054074) that this isn't python 3.5 compatible, I don't think you have to worry. The only parts that aren't compatible are the `.pyi` files, but that's ok, they're only used by MyPy. I've tested running `lint-python.sh` under python 3.5 and it doesn't give any errors. (MyPy does, however, get confused and tell me that a couple lines are wrong when they aren't. It doesn't seem related though).

Thanks for taking the time to review the code! 

When I run the linter file (lint-python.py), Mypy doesn't catch any errors, however, when I run mypy from the command line as such:
`MYPYPATH=/path/to/stub/files mypy test/functional/contains_mypy_errors.py` 
incompatible type errors are properly caught and displayed. I haven't been able to find too much on this unexpected behavior from Mypy other than this Stackoverflow post: https://stackoverflow.com/questions/55437746/how-do-i-correctly-set-mypypath-to-pick-up-stubs-for-mypy.

**Comment by adaminsky** [comment]
I'm new to mypy, but I found this [issue](https://github.com/python/mypy/issues/1645) which seems to address the problem that was previously encountered. It seems like the `--namespace-packages` flag will fix the issue, but it may be easier to just add the empty `__init__.py` file to `test/functional/data` which fixes the problem too.

To test this,  I included the `__init__.py` file in the `data` directory and then `mypy test/functional` runs without any errors. Creating an obvious type error inside the `data` package also results in the error being correctly detected. This will make the generated stub files no longer necessary, so the type annotations can be made directly in the source files.

**Comment by troygiorshev** [comment]
@adaminsky thanks for looking at this!  Great to see that we might not need the stub files after all!  I don't think we're all the way there yet thought, creating `test/functional/data/__init__.py` is already the second commit of this PR :)

Did you try making a type error inside a test in `test/functional/`?  I still can't get those to be caught.

**Comment by troygiorshev** [comment]
I've tried changing MyPy's configuration options, maybe that will do the trick.  To try this too, create a config file, say, `test/functional/data/mypy.ini` with the following in it:

```
[mypy]
check_untyped_defs = True
no_implicit_optional = True
```

Then running MyPy with `mypy --config-file test/functional/data/mypy.ini test/functional` gives tonnes of errors (finally 🎉) but I'm not sure that all of them are useful yet.

**Comment by adaminsky** [comment]
@troygiorshev I mislabeled the types in the `BaseNode.__init__` function in `test/functional/example_test.py` and when I run `mypy test/functional`, it gives an error. I tried stuff like `1 + 'test'` to create type errors, but that doesn't work because Mypy assumes everything is untyped unless a function is fully annotated ([Mypy doc](https://mypy.readthedocs.io/en/stable/common_issues.html#no-errors-reported-for-obviously-wrong-code)). I think this is why running with your config file finds so many errors because the `check_untyped_defs` flag makes Mypy treat everything as typed.

**Comment by laanwj** [comment]
I understand mypy is desirable for type checking, but I see a potential maintainability issue with the stub files: many changes to the test framework will now require updating these files as well. I'm not sure we should be putting this burden on every contributor.

File: test/functional/test_framework/test_framework.py (hunk 1 above)
Comment (kiminuo) [reply test/functional/test_framework/test_framework.py:717]: > Unfortunately not. We need the type:ignore or else MyPy fails completely. I've tried things like putting them both on the same line but it looks like flake8 and MyPy just don't play nicely together in this sort of conflict.

Just to clarify: https://mypy.readthedocs.io/en/stable/common_issues.html#silencing-linters - this does not work? Is it what you have tried?

**Comment by fanquake** [comment]
@Ghorbanian thanks for your work here, but I agree with @laanwj in regards to the maintenance overhead of adding these stubs. Since this PR was opened, there have been upstream changes which have negated the needs for some of the commits here, such as ignoring the zmq imports (the pyzmq package now comes with stubs).

I'm going to close this PR, and I think we can renew discussion in #19389, and potentially make some other mypy related changes in the mean time.

**Pull request closed without merging**